In [12]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

In [11]:
# 载入数据集
mnist = input_data.read_data_sets('MNIST_data',one_hot=True)

# 每个批次的大小
batch_size = 100
# 一共有多少个批次
n_batch = mnist.train.num_examples // batch_size

# 定义两个变量
x = tf.placeholder(tf.float32,[None,784])
y = tf.placeholder(tf.float32,[None,10])

# 创建一个简单的神经网络
W = tf.Variable(tf.zeros([784,10]))
b = tf.Variable(tf.zeros([10]))
prediction = tf.nn.softmax(tf.matmul(x,W)+b)

# 二次代价函数
# loss = tf.reduce_mean(tf.square(y-prediction))
# 交叉熵
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y,logits=prediction))
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(loss)

# 结果存放
correct_prediction = tf.equal(tf.argmax(y,1),tf.argmax(prediction,1))
# 准确率
accuracy = tf.reduce_mean(tf.cast(correct_prediction,tf.float32))

# 初始化
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
#     for _ in range(200):
#         print(sess.run([w,b]))
    for epoch in range(21):
        for batch in range(n_batch):
            batch_xs,batch_ys = mnist.train.next_batch(batch_size)
            sess.run(train_step,feed_dict={x:batch_xs,y:batch_ys})
        
        acc = sess.run(accuracy,feed_dict={x:mnist.test.images,y:mnist.test.labels})
        print("Iter" + str(epoch) + ",Testing Accuracy " + str(acc)) 

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz
Iter0,Testing Accuracy 0.8946
Iter1,Testing Accuracy 0.9093
Iter2,Testing Accuracy 0.9128
Iter3,Testing Accuracy 0.9162
Iter4,Testing Accuracy 0.9188
Iter5,Testing Accuracy 0.918
Iter6,Testing Accuracy 0.9182
Iter7,Testing Accuracy 0.9208
Iter8,Testing Accuracy 0.9224
Iter9,Testing Accuracy 0.9219
Iter10,Testing Accuracy 0.9232
Iter11,Testing Accuracy 0.9235
Iter12,Testing Accuracy 0.9231
Iter13,Testing Accuracy 0.9236
Iter14,Testing Accuracy 0.9252
Iter15,Testing Accuracy 0.9244
Iter16,Testing Accuracy 0.9268
Iter17,Testing Accuracy 0.9252
Iter18,Testing Accuracy 0.9251
Iter19,Testing Accuracy 0.9262
Iter20,Testing Accuracy 0.9272


# 二次代价函数

Iter0,Testing Accuracy 0.8763
Iter1,Testing Accuracy 0.8931
Iter2,Testing Accuracy 0.8994
Iter3,Testing Accuracy 0.9039
Iter4,Testing Accuracy 0.9064
Iter5,Testing Accuracy 0.9106
Iter6,Testing Accuracy 0.9121
Iter7,Testing Accuracy 0.9127
Iter8,Testing Accuracy 0.9153
Iter9,Testing Accuracy 0.9163
Iter10,Testing Accuracy 0.9173
Iter11,Testing Accuracy 0.9179
Iter12,Testing Accuracy 0.9186
Iter13,Testing Accuracy 0.9187
Iter14,Testing Accuracy 0.9195
Iter15,Testing Accuracy 0.9202
Iter16,Testing Accuracy 0.9201
Iter17,Testing Accuracy 0.9202
Iter18,Testing Accuracy 0.9206
Iter19,Testing Accuracy 0.9208
Iter20,Testing Accuracy 0.9214

# cross_entropy
Iter0,Testing Accuracy 0.8946
Iter1,Testing Accuracy 0.9093
Iter2,Testing Accuracy 0.9128
Iter3,Testing Accuracy 0.9162
Iter4,Testing Accuracy 0.9188
Iter5,Testing Accuracy 0.918
Iter6,Testing Accuracy 0.9182
Iter7,Testing Accuracy 0.9208
Iter8,Testing Accuracy 0.9224
Iter9,Testing Accuracy 0.9219
Iter10,Testing Accuracy 0.9232
Iter11,Testing Accuracy 0.9235
Iter12,Testing Accuracy 0.9231
Iter13,Testing Accuracy 0.9236
Iter14,Testing Accuracy 0.9252
Iter15,Testing Accuracy 0.9244
Iter16,Testing Accuracy 0.9268
Iter17,Testing Accuracy 0.9252
Iter18,Testing Accuracy 0.9251
Iter19,Testing Accuracy 0.9262
Iter20,Testing Accuracy 0.9272